src/main/java/com/thevoxelbox/voxelsniper/SnipeData.java

```diff
@@ -117,108 +124,89 @@ public void setVoxelHeight(int voxelHeight) {
         this.voxelList = voxelList;
     }
 
     public Message getVoxelMessage() {
-        return this.voxelMessage;
+        return voxelMessage;
     }
 
     public void setVoxelMessage(Message voxelMessage) {
         this.voxelMessage = voxelMessage;
     }
 
     public World getWorld() {
-        return this.owner.getPlayer().getWorld();
+        return owner.getPlayer().getWorld();
     }
 
     public boolean isLightningEnabled() {
-        return this.lightning;
+        return lightning;
     }
 
     public void setLightningEnabled(boolean lightning) {
         this.lightning = lightning;
     }
 
     public boolean isRanged() {
-        return this.ranged;
+        return ranged;
     }
 
     public void setRanged(boolean ranged) {
         this.ranged = ranged;
     }
 
-    public Key<? extends BaseValue<?>> getVoxelInkKey() {
-        return this.voxelInkKey;
-    }
-
-    public Object getVoxelInkValue() {
-        return this.voxelInkValue;
-    }
-
-    public <V extends BaseValue<T>, T> void setVoxelInk(Key<V> key, T value) {
-        this.voxelInkKey = key;
-        this.voxelInkValue = value;
+    public Map<BlockTrait<?>, Object> getVoxelInkTraits() {
+        return Collections.unmodifiableMap(voxelInkTraits);
     }
 
-    public Key<? extends BaseValue<?>> getReplaceInkKey() {
-        return this.replaceInkKey;
+    // Attempts to add the traitValues as the current placement ink.  If the trait values can't be used with the current
+    // voxel state, no changes are made.
+    public void setVoxelInkTraits(Map<BlockTrait<?>, Object> traitValues) {
+        Optional<BlockState> optState = BlockHelper.addTraits(voxelState.getType().getDefaultState(), traitValues);
+        if (optState.isPresent()) {
+            voxelState = optState.get();
+            voxelInkTraits.clear();
+            voxelInkTraits.putAll(traitValues);
+        }
     }
 
-    public Object getReplaceInkValue() {
-        return this.replaceInkValue;
+    public Map<BlockTrait<?>, Object> getReplaceInkTraits() {
+        return Collections.unmodifiableMap(replaceInkTraits);
     }
 
-    public <V extends BaseValue<T>, T> void setReplaceInk(Key<V> key, T value) {
-        this.replaceInkKey = key;
-        this.replaceInkValue = value;
+    // Attempts to add the traitValues as the current replacement ink.  If the trait values can't be used with the
+    // current replacement state, no changes are made.
+    public void setReplaceInkTraits(Map<BlockTrait<?>, Object> traitValues) {
+        Optional<BlockState> optState = BlockHelper.addTraits(replaceState.getType().getDefaultState(), traitValues);
+        if (optState.isPresent()) {
+            replaceState = optState.get();
+            replaceInkTraits.clear();
+            replaceInkTraits.putAll(traitValues);
+        }
     }
 
     public Sniper owner() {
         return this.owner;
     }
 
-    /**
-     * Reset to default values.
-     */
-    public void reset() {

```

Comment:
Is Reset no longer a thing?

Reply:
It was not being called anywhere so I removed it.